### ISSUE TITLE
Error while proxying Websocket connection 

### ISSUE THREAD

**Issue by nmarcetic** [vernemq/vernemq#1457]
### Environment

- VerneMQ Version: Latest 
- OS: Official docker image
- Erlang/OTP version (if building from source): Official docker image
- VerneMQ configuration (vernemq.conf) or the changes from the default: No changes
- Cluster size/standalone: standalone

### Expected behavior
Proxying Websocket's 
- Please describe the expected behaviour.
I am working on [mproxy](https://github.com/mainflux/mproxy), where VerneMQ is our default broker. Proxying works perfectly with MQTT but not for WS. 
It works perfect with other brokers like [Aedes](https://github.com/moscajs/aedes) or [EMQX](https://www.emqx.io/products/broker). 

I understand that my issue can be caused by my implementation etc... and I am not expecting you to dig into my code or nothing similar. What I am expecting is to check out my VerneMQ error log and maybe point me in the right direction or just think what can cause this error. Thanks!
 
### Actual behaviour
Looks like ws handler crashs.
```
12:11:37.289 [info] Try to start vmq_plumtree: ok
12:11:37.781 [info] Try to start vmq_generic_msg_store: ok
12:11:38.019 [info] loaded 0 subscriptions into vmq_reg_trie
12:11:38.024 [info] cluster event handler 'vmq_cluster' registered
12:12:02.488 [error] CRASH REPORT Process <0.453.0> with 0 neighbours crashed with reason: bad record state in vmq_websocket:add_socket/2 line 194
12:12:02.488 [error] Ranch listener {{172,17,0,2},8080} terminated with reason: bad record state in vmq_websocket:add_socket/2 line 194

```
- Please include minimal steps to reproduce the issue, if possible.
It's too complicated. Just to try to analyze the logs.

**Comment by ioolkos** [comment]
Does a normal WebSocket connection work? Can you post snippet of your config, regarding the WS listener?

**Comment by nmarcetic** [comment]
Yes, the normal connection works from test client, which is paho go lib (e.g without proxy, connecting directly to Vernemq). 
`Can you post snippet of your config, regarding the WS listener?` 
I didn't configure listener or any custom config. I use official docker image, running broker with 
```
 docker run -d -p 1883:1883 -p 8080:8080 --name vernemq-native -e "DOCKER_VERNEMQ_ACCEPT_EULA=yes" -e "DOCKER_VERNEMQ_ALLOW_ANONYMOUS=on" vernemq/vernemq:latest

```

**Comment by ioolkos** [comment]
Ok, then there should be a WS (not WSS!) listener running on 8080.
Strange. What would be the difference from Paho Go client vs connecting from MProxy?

**Comment by nmarcetic** [comment]
There is a lot of difference, with mproxy client (e.g paho lib) connects to mproxy WS server (server do connection Upgrade) and this server holds connection (its lower level, pure WS). Then server connections to VerneMQ also raw WS and start proxying by Read/Write both connections (message sent from a client will be written to a connection that mproxy holds with server). 
It works perfectly for raw TCP (e.g MQTT proxying) and the same concept works with WS, but another broker, not VerneMQ. So I don't get it what is different actually. Does this error log say something that makes sense to you? What it means 
`terminated with reason: bad record state in vmq_websocket:add_socket` what is a record state?

**Comment by nmarcetic** [comment]
Just to add, in this proxying flow ^ error occurs on CONNECT request. So I successfully create a tunnel (connect to VerneMQ with raw WS and tunnel client connection) but on the first CONNECT packet WS listener crash. Yes pure WS not WSS.

**Comment by ioolkos** [comment]
Can you wireshark all of this handshake?

The state record represents the current state of this module: https://github.com/vernemq/vernemq/blob/baa9b2794a1a50980c80a90d1ba51c6e84716f14/apps/vmq_server/src/vmq_websocket.erl#L29

It makes a copy of the state and returns a new state with the Socket added here. The error must be happening while copying state.
https://github.com/vernemq/vernemq/blob/baa9b2794a1a50980c80a90d1ba51c6e84716f14/apps/vmq_server/src/vmq_websocket.erl#L193

**Comment by nmarcetic** [comment]
Ok, let me try to investigate handshake a bit more. I Hope I will get back with more info. Thanks  @ioolkos

**Comment by nmarcetic** [comment]
@ioolkos In this flow of proxying, I was missing `Sec-Webscoket-Protocol` header set to `mqtt` like [paho is doing with gorila WebSocket library](https://github.com/eclipse/paho.mqtt.golang/blob/master/websocket.go#L25)  when I am connecting to VerneMQ from proxy. I noticed that Mosquitto broker requires this also (some brokers not, that's why it worked with Aedes and Emqx).
Thanks to your idea with Wireshark, I realized this (compering success handshake client -> broker with failing one client -> mproxy  -> broker). 

Now everything works as expected. Thanks you so much for your support and prompt replays. Really appreciate!
